### PR TITLE
IBM Public Cloud Support

### PR DESCRIPTION
**Pull request by csrwng**

### CODE REVIEW

**Comment by csrwng** [comment]
@derekwaynecarr

**Comment by derekwaynecarr** [comment]
/assign @derekwaynecarr 

FYI @smarterclayton @eparis

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
+---
+title: IBM Public Cloud Support
+authors:
+  - "@csrwng"
+reviewers:
+  - "@derekwaynecarr"
```
Comment (sttts) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: expecting the owners of the control plane operators here, compare https://github.com/openshift/enhancements/pull/202/files#diff-8d49ecd990d312a72c0ffcdd1784ad05R92.

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+that applies these manifests and instantiates worker nodes that add themselves to 
+the new cluster. Minting of kubelet certificates for these worker nodes is handled
+by IBM automation.
+
+Components that run on the management cluster include:
+- etcd\+
+- kubernetes apiserver
+- kubernetes controller manager
+- kubernetes scheduler
+- openshift apiserver
```
Comment (sttts) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: has it been discussed what is needed to move that into the customer cluster?

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+Components and changes required for this phase:
+
+#### Control Plane Operators
+Break up the single Beta control plane operator into different operators that
+vendor code from the self-hosted control plane operators in the product. The vendored
+code should include config observers that assemble a new configuration for their
+respective control plane components. This will ensure that drift in future versions
+is kept under control and that a single code base is used to manage control plane
+configuration.
```
Comment (sttts) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: where is this beta control plane operator?

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+Enable OpenShift 4.x deployment on IBM Cloud.
+
+### Goals
+
+- Enable the deployment of OpenShift 4.x on IBM Public Cloud by providing the IBM
+  Public Cloud team the necessary tools to generate manifests needed for a hosted
+  control plane.
+- Ensure that this deployment model remains functional through regular e2e testing
+  on IBM Public Cloud.
```
Comment (sttts) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: "regular" means we get a normal CI job in the openshift org making sure our control plane code changes don't break it?

Comment (sttts) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: Will the CI job be blocking for OpenShift PRs? E.g. if deployment topology changes and some apiserver suddenly does not serve certain APIs because they moves, their CI will break.

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md (hunk 1 above)
Comment (csrwng) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: @sttts did you mean a different pr? the link above is for this one

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md (hunk 2 above)
Comment (csrwng) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: It hasn't... when we first looked at this, however, we were in a catch 22 situation, where in order to be able to schedule pods we needed the openshift crds and controller functional.

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md (hunk 3 above)
Comment (csrwng) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: we're working on it, will be added to the current hypershift-toolkit repo. For the second phase we will create separate repos for each of the control plane controllers.

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md (hunk 4 above)
Comment (csrwng) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: > "regular" means we get a normal CI job in the openshift org making sure our control plane code changes don't break it?

That is the plan. We currently have a periodic job that creates clusters based on the 4.3 branches. One will be added for the master/4.4 branch.


> Will the CI job be blocking for OpenShift PRs? E.g. if deployment topology changes and some apiserver suddenly does not serve certain APIs because they moves, their CI will break.

That will be harder. We would require capacity on the IBM Cloud to run that many jobs. Not sure that is feasible right now. The periodic job should block a release, but not individual PRs

Comment (sttts) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: We describe those apiserver changes (they are happening, now for oauth) in enhancements. The IBM team has to watch that repo to be informed.

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md (hunk 1 above)
Comment (sttts) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: No, meant this one. Just expected the owners of the mentioned operators to be informed about these plans by being reviewer of the enhancement.

Comment (csrwng) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: Ack, will add more reviewers

**Comment by csrwng** [comment]
@deads2k @mfojtik @ironcladlou @spadgett @abhinavdahiya @crawford @miabbott 

Please let me know if I should include other reviewers

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+The cluster version operator running on the management cluster connects to the Kubernetes
+API server via the service network and starts applying manifests for SLOs that must run
+on the worker nodes. These include (among others):
+
+- network (dns, multus)
+- ingress
+- image registry
+- OLM
+- samples
+- console
```
Comment (spadgett) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: The console node selector currently matches only master nodes. There is no technical reason the console can't run on workers, just that we wanted to avoid running it next to user workloads if possible since the console backend handles all logged-in user tokens. This will likely have to change, though.

cc @benjaminapetersen

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+present, should include the annotation to exclude them. In most cases, only the
+minimum set of manifests to allow skipping the component should be annotated.
+However, in the case of the Machine API and Machine Configuration operators, 
+the CRDs that represent machines, machinesets and autoscalers should also be
+skipped. Monitoring alerts for components that do not get installed in the user
+cluster should also be skipped where possible.
+
+#### Console Changes
+The console should not report the control plane as being down if no metrics
+datapoints exist for control plane components in this configuration.
```
Comment (spadgett) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: Ack, this shouldn't be an issue.

cc @rawagner @andybraren

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+Components that run on the management cluster include:
+- etcd\+
+- kubernetes apiserver
+- kubernetes controller manager
+- kubernetes scheduler
+- openshift apiserver
+- openshift controller manager
+- cluster version operator
+- control plane operator(s)\*
+- oauth server\+
```
Comment (spadgett) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: @csrwng Are there any lingering issues where the console backend rejects the OAuth server certificate in this deployment?

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md (hunk 6 above)
Comment (andybraren) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: I believe Control Plane components would be shown as "Not available" if that's acceptable (better than "Down").

image

Comment (spadgett) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: @andybraren We probably need to remove that since we never expect to have control plane metrics. It's misleading to say not available.

Comment (rawagner) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: Agreed, created an issue to track this for 4.5 Dashboards https://issues.redhat.com/browse/KNIP-1252

**Comment by miabbott** [comment]
cc: @lucab

**Comment by lucab** [comment]
From an OS point of view, there are at least two things in this proposal that looks uncomfortably hairy to me:
 1. There is no mention of which cloud flavor this is trying to target. At a compute level, "IBM Cloud" is really an umbrella label for three different kinds of infrastructure: Classic, Gen 1, and Gen 2.
Out of those three, only the latter (IMHO) qualifies as a proper environment where we can sanely support provisioning "cattle nodes" with RHCOS and Ignition.
 1. This briefly mentions several topics which seem to require host-level customization (e.g. VPN setup, service network, certificate minting, IBM-specific automation) without ever mentioning how the logic for that is containerized and provided to the nodes at provisioning time (i.e. before kubelet is bootstrapped).

In short, the post-GA story is under-specified so it's quite hard to judge it. The rest of the document seems to hint at a heavy UPI+RHEL environment, which offers a lot of escape hatches and is more likely to result in a "pet nodes" provisioning flow.

If that's indeed the priority, then it would be better to descope RHCOS and leave if for "future exploration" (with the risk that it may be very hard or impossible to retrofit). If RHCOS workers are instead a requirement, then clarifying the points above may result in vastly different design and required work.

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+  on IBM Public Cloud.
+- Make the necessary product changes to make IBM Public Cloud a fully supported 
+  cloud provider.
+- Ensure that clusters running on IBM Public Cloud are conformant OpenShift clusters
+  (pass Kubernetes and OpenShift conformance e2e tests). Fully document any tests that
+  must be skipped due to differences from a traditional deployment.
+
+### Non-Goals
+
+- Make hosted control planes a supported deployment model outside of IBM Public Cloud.
```
Comment (cgwalters) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: I find this confusing...IBM Public Cloud means a lot of different things, and a sort of baseline obvious one is having OpenShift support the default "self driving" path in their existing IaaS.  But I guess we're doing hosted control plane *first*?

Maybe the enhancement should be called: "IBM Public Cloud Hosted Control Plane" ?

And one thing I would say here is that we should think of this "fairly" - if some other IaaS showed up and was willing to commit significant resources to maintaining a similar thing...  clearly vast amounts of the design would likely be shared.  But that can come later.

Comment (csrwng) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: Absolutely, I think the things we learn from this work is something we can likely reuse in other cases. And perhaps at the proper time, make the pattern something standalone that's configured per provider. So definitely, this non-goal is a point-in-time statement.

Comment (csrwng) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: > I find this confusing...IBM Public Cloud means a lot of different things, and a sort of baseline obvious one is having OpenShift support the default "self driving" path in their existing IaaS. But I guess we're doing hosted control plane first?

At least in the foreseeable future, supporting the self-hosted path is not a priority afaik, but @derekwaynecarr can likely provide more insight into that.

**Comment by csrwng** [comment]
> In short, the post-GA story is under-specified so it's quite hard to judge it. The rest of the document seems to hint at a heavy UPI+RHEL environment, which offers a lot of escape hatches and is more likely to result in a "pet nodes" provisioning flow.

@lucab thank you for the feedback. Yes, this proposal is definitely under-specified where it comes to RHCOS. It's more a statement that we don't expect to continue supporting IBM cloud without RHCOS forever. We should have a separate enhancement proposal/design specifically for RHCOS, given that as I understand it, the RHCOS team has already done some initial investigations around this.

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md (hunk 7 above)
Comment (csrwng) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md]: @spadgett no issue at the moment. Thx!

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+2. GA: In-tree support for ingress, image registry, storage, auth, load balancers.
+3. Post-GA: Add support for RHCOS workers
+
+In more detail:
+
+### Beta
+
+Enables an OpenShift cluster to be hosted on top of a Kubernetes/OpenShift cluster. 
+
+Given a release image, a CLI tool generates manifests that instantiate the control plane
```
Comment (cgwalters) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md:94]: OS upgrades for the worker nodes is owned by the customer?  Or does IBM provide tooling for that?  Are they using [openshift-ansible](https://github.com/openshift/openshift-ansible/)?

In the "Post-GA" world with RHCOS...do we forsee trying to enable the MCO to manage upgrades for the workers w/RHCOS?

Comment (wking) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md:94]: Isn't this covered in the _Managed Workers_ section below?

Comment (cgwalters) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md:94]: The second half is yes, thanks!

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+Support IBM public cloud as a provider.
+
+### Post-GA
+
+Add support for managed RHCOS nodes.
+
+#### Managed Workers
+RHCOS adds support for bootstrapping on IBM Public Cloud. The MCO is added to 
+the components that get installed on the management cluster.  This enables upgrading 
+of RHCOS nodes using the same mechanisms as in self-hosted OpenShift.
```
Comment (wking) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md:237]: So compute nodes still run machine-config daemons and cluster admins can write MachineConfig entries, create MachineSets, and all that good stuff?  They just don't have any objects representing or control over the control-plane machines?

**Comment by derekwaynecarr** [comment]
The service is now GA and running 4.4.11, so let's merge this and then make updates to explain usage of Cluster Profile(s).

/approve
/lgtm

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+`internal-openshift-hosted`
+
+#### Excluded Manifests
+Manifests for the components that should be skipped by the CVO whenever
+the `EXCLUDE_MANIFESTS=internal-openshift-hosted` environment variable is 
+present, should include the annotation to exclude them. In most cases, only the
+minimum set of manifests to allow skipping the component should be annotated.
+However, in the case of the Machine API and Machine Configuration operators, 
+the CRDs that represent machines, machinesets and autoscalers should also be
+skipped. Monitoring alerts for components that do not get installed in the user
```
Comment (lilic) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md:188]: This part won't be that easy, unless we get provided with a list of it. Also it will not fire if there are no metrics for those components, so don't think its a problem.

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+management cluster.
+
+Supporting this type of deployment in OpenShift 4.x requires a different approach to
+installing OpenShift and managing its control plane. Instead of bootstrapping a cluster 
+control plane from scratch on virtual machines, a cluster control plane can be created 
+by applying a set of manifests to an existing cluster in a new namespace. 
+
+Changes are required in different areas of the product in order to make clusters deployed 
+using this method viable. These include changes to the cluster version operator (CVO), web 
+console, second level operators (SLOs) deployed by the CVO, and RHCOS.
```
Comment (s-urbaniak) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md:51]: please don't use SLO as an acronym here. It is very commonly used for "Service Level Objectives".

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+
+Enables an OpenShift cluster to be hosted on top of a Kubernetes/OpenShift cluster. 
+
+Given a release image, a CLI tool generates manifests that instantiate the control plane
+components so that they can be applied to a namespace. IBM provides the automation
+that applies these manifests and instantiates worker nodes that add themselves to 
+the new cluster. Minting of kubelet certificates for these worker nodes is handled
+by IBM automation.
+
+Components that run on the management cluster include:
```
Comment (s-urbaniak) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md:100]: for cluster-monitoring: as apiserver monitoring is effectively being disabled when running in "ROKS mode" what is monitoring control plane components on the management cluster?

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+the `EXCLUDE_MANIFESTS=internal-openshift-hosted` environment variable is 
+present, should include the annotation to exclude them. In most cases, only the
+minimum set of manifests to allow skipping the component should be annotated.
+However, in the case of the Machine API and Machine Configuration operators, 
+the CRDs that represent machines, machinesets and autoscalers should also be
+skipped. Monitoring alerts for components that do not get installed in the user
+cluster should also be skipped where possible.
+
+#### Console Changes
+The console should not report the control plane as being down if no metrics
```
Comment (s-urbaniak) [inline enhancements/ibm-public-cloud/ibm-public-cloud-support.md:192]: Are all edge cases for disabling monitoring of control plane components on the worker clusters covered in https://github.com/openshift/cluster-monitoring-operator/pull/705 ?

Comment (csrwng) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md:192]: So far yes

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md (hunk 11 above)
Comment (csrwng) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md:188]: This is what we addressed with https://github.com/openshift/cluster-monitoring-operator/pull/705
No other alerts related to control plane components have surfaced.

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md (hunk 12 above)
Comment (csrwng) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md:51]: ack, will do a follow-up to remove.

File: enhancements/ibm-public-cloud/ibm-public-cloud-support.md (hunk 13 above)
Comment (csrwng) [reply enhancements/ibm-public-cloud/ibm-public-cloud-support.md:100]: IBM is running their own monitoring solution on their management/tugboat clusters.
